fcrepo-clients/src/main/java/edu/unc/lib/dl/fcrepo4/TransactionalFcrepoClient.java

```diff
@@ -101,15 +108,33 @@ public FcrepoResponse executeRequest(URI uri, HttpRequestBase request)
             if (needsRequestBodyRewrite(request)) {
                 rewriteRequestBodyUris(request);
             }
-            URI requestUri = uri;
             if (!uri.toString().contains("tx:")) {
-                requestUri = rewriteUri(uri);
+                requestUri = rewriteUri(requestUri);
                 request.setURI(requestUri);
             }
             FcrepoResponse resp = super.executeRequest(requestUri, request);
             return rewriteResponseBodyUris(resp);
         }
-        return super.executeRequest(uri, request);
+        return super.executeRequest(requestUri, request);
+    }
+
+    /**
+     * Rebase fedora URI to the domain expected by this client.
+     *
+     * @param uri
+     * @return
+     */
+    private URI rebaseUri(URI uri) {
+        URI requestUri = uri;
+        String uriString = uri.toString();
+        if (!uriString.startsWith(baseUri)) {
+            int index = uriString.indexOf(REST);
+            if (index == -1) {
+                throw new IllegalArgumentException("Requested URI not within allowed domain");
+            }
+            requestUri = URI.create(URIUtil.join(baseUri, uriString.substring(index + 4)));
```

Comment:
would you mind changing the numeral to something like REST_LENGTH?

Reply:
updated